lib/Watcher.js

```diff
@@ -0,0 +1,184 @@
+        (error: Error) => {
+          if (error) {
+            reject(error);
+            return;
+          }
+          resolve();
+        });
+
+      client.on('subscription', (resp: Object) => {
+        if (resp.subscription !== SUBSCRIPTION_NAME) {
+          return;
+        }
+
+        const added = [];
+        const removed = [];
+        resp.files.forEach((file: Object) => {
+          const normalizedPath = normalizePath(file.name, this.workingDirectory);
+          if (normalizedPath.indexOf('/node_modules/') !== -1) {
+            return;
+          }
+          if (this.excludes.some((pattern: string): boolean =>
+            minimatch(normalizedPath, pattern))) {
+            return;
+          }
+          if (file.exists) {
+            added.push({ path: normalizedPath, mtime: +file.mtime_ms });
+          } else {
+            removed.push({ path: normalizedPath });
+          }
+        });
+        if (added.length) {
+          this.onFilesAdded(added);
+        }
+        if (removed.length) {
+          this.onFilesRemoved(removed);
+        }
+      });
+    });
+  }
+
+  startSubscription({ client }: { client: fbWatchman.Client }): Promise<void> {
+    return new Promise((resolve: Function, reject: Function) => {
+      client.command(['watch-project', this.workingDirectory], (error: Error, resp: Object) => {
+        if (error) {
+          reject(error);
+          return;
+        }
+
+        if ('warning' in resp) {
+          loglevel.warn(`WARNING received during watchman init: ${resp.warning}`);
+        }
+
+        this.subscribe({
+          client,
+          fbWatch: resp.watch,
+          relativePath: resp.relative_path,
+        }).then(resolve).catch(reject);
+      });
+    });
+  }
+
+  initialize(): Promise<void> {
+    return new Promise((resolve: Function, reject: Function) => {
+      this.initializeWatchman().then(resolve).catch(() => {
+        loglevel.info(
+          "Couldn't initialize watchman watcher. Falling back to polling.");
+        this.initializePolling().then(resolve).catch(reject);
+      });
+    });
+  }
+
+  /**
+   * Get all files from the watchman-powered cache. Returns a promise that will
+   * resolve if watchman is available, and the file cache is enabled. Will
+   * resolve immediately if previously initialized.
+   */
+  initializeWatchman(): Promise<void> {
+    return new Promise((resolve: Function, reject: Function) => {
+      const client = new fbWatchman.Client();
+      client.on('error', (error: Error) => {
+        reject(error);
+      });
+      client.capabilityCheck({
+        optional: [],
+        required: ['relative_root'],
+      }, (error: Error) => {
+        if (error) {
+          client.end();
+          reject(error);
+        } else {
+          this.startSubscription({ client }).then(resolve).catch(reject);
+        }
+      });
+    });
+  }
+
+  initializePolling(): Promise<void> {
+    setInterval(() => {
```

Comment:
That makes sense to me. It might be worth considering making this pluggable, so you can use watchman or swap it out for some other thing if you want (e.g. polling, fsevents, or some other file watcher).